File: src/BipsPl/top_level.pl
/*-------------------------------------------------------------------------*
 * GNU Prolog                                                              *
 *                                                                         *
 * Part  : Prolog buit-in predicates                                       *
 * File  : top_level.pl                                                    *
 * Descr.: top Level                                                       *
 * Author: Daniel Diaz                                                     *
 *                                                                         *
 * Copyright (C) 1999-2013 Daniel Diaz                                     *
 *                                                                         *
 * This file is part of GNU Prolog                                         *
 *                                                                         *
 * GNU Prolog is free software: you can redistribute it and/or             *
 * modify it under the terms of either:                                    *
 *                                                                         *
 *   - the GNU Lesser General Public License as published by the Free      *
 *     Software Foundation; either version 3 of the License, or (at your   *
 *     option) any later version.                                          *
 *                                                                         *
 * or                                                                      *
 *                                                                         *
 *   - the GNU General Public License as published by the Free             *
 *     Software Foundation; either version 2 of the License, or (at your   *
 *     option) any later version.                                          *
 *                                                                         *
 * or both in parallel, as here.                                           *
 *                                                                         *
 * GNU Prolog is distributed in the hope that it will be useful,           *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU       *
 * General Public License for more details.                                *
 *                                                                         *
 * You should have received copies of the GNU General Public License and   *
 * the GNU Lesser General Public License along with this program.  If      *
 * not, see http://www.gnu.org/licenses/.                                  *
 *-------------------------------------------------------------------------*/


:-	built_in.

:-	ensure_linked([consult / 1, load / 1]).


top_level :-
	current_prolog_flag(prolog_name, Name),
	current_prolog_flag(prolog_version, Version),
	current_prolog_flag(prolog_copyright, Copyright),
	current_prolog_flag(address_bits, Bits),
	current_prolog_flag(compiled_at, Date),
	current_prolog_flag(c_cc, CC),
	format(top_level_output, '~N~a ~a (~d bits)~n', [Name, Version, Bits]),
	format(top_level_output, 'Compiled ~a with ~a~n', [Date, CC]),
	write(top_level_output, 'By Daniel Diaz'),
	nl,
	format(top_level_output, '~a~n', [Copyright]),
	break.




break :-
	'$call_c'('Pl_Set_Ctrl_C_Handler_0'),
	'$sys_var_read'(10, Level),
	'$sys_var_read'(11, B),
	g_read('$all_solutions', All),
	(   Level > 0 ->
	    format(top_level_output, '~N{Break Level ~d}~n', [Level])
	;   true
	),
	'$sys_var_inc'(10),
	g_read('$cmd_line_consult_file', LFile),
	'$exec_cmd_line_consult_files'(LFile),
	g_read('$cmd_line_entry_goal', LGoal),
	'$exec_cmd_line_entry_goals'(LGoal),
	g_assign('$cmd_line_entry_goal', []),
	'$top_level1',
	'$sys_var_dec'(10),
	'$sys_var_write'(11, B),
	g_assign('$all_solutions', All),
	(   Level > 0 ->
	    format(top_level_output, '~N{End Break}~n', [Level])
	;   true
	).




'$top_level1' :-
	repeat,
	'$catch_internal'('$top_level2', X, '$top_level_exception'(X), false), !.




'$top_level_abort' :-
	'$reinit_after_exception',
	'$sys_var_read'(11, B),
	write(top_level_output, 'execution aborted\n'),
	'$catch_sync_for_fail_at'(B).




'$top_level_stop' :-
	'$reinit_after_exception',
	'$sys_var_read'(11, B),
	'$catch_sync_for_fail_at'(B).




'$top_level_exception'('$post_query_exception'(X)) :-
	'$reinit_after_exception',
	!,
	format(top_level_output, '~Ntop-level exception: ', []),
	write_term(top_level_output, X, [quoted(true), numbervars(false), namevars(false)]),
	nl(top_level_output),
	fail.
	

'$top_level_exception'(X) :-
	'$reinit_after_exception',
	format(top_level_output, '~Nuncaught exception: ', []),
	write_term(top_level_output, X, [quoted(true), numbervars(false), namevars(false)]),
	nl(top_level_output),
	fail.



'$reinit_after_exception' :-
% g_read('$char_conv', CharConv),
% set_prolog_flag(char_conversion, CharConv),
	(   '$sys_var_read'(12, 1) ->
	    g_read('$user_prompt', UserPrompt),
	    '$set_linedit_prompt'(UserPrompt)
	;   true
	).




'$top_level2' :-
	repeat,
	'$get_current_B'(B),   % the current choice-point
%	'$sys_var_read'(7, B), % the last Handler created by catch/3 (what is better ???)
	'$sys_var_write'(11, B),
%write('top-level catcher'(B)), nl,
	'$write_indicator',
% current_prolog_flag(char_conversion, CharConv),
% g_assign('$char_conv', CharConv),
% set_prolog_flag(char_conversion, off),
	Prompt = '| ?- ',
	(   '$sys_var_read'(12, 1) ->
	    '$get_linedit_prompt'(UserPrompt),
	    g_assign('$user_prompt', UserPrompt),
	    '$set_linedit_prompt'(Prompt)
	;   write(top_level_output, Prompt)
	),
	flush_output(top_level_output),
	'$read_query'(X, QueryVars),
	(   '$sys_var_read'(12, 1) ->
	    '$set_linedit_prompt'(UserPrompt)
	;   true
	),
% set_prolog_flag(char_conversion, CharConv),
	sort(QueryVars, QueryVars1),
	(   X == end_of_file ->
	    nl(top_level_output), !
	;   user_time(Time0),
	    (   '$exec_query'(X, QueryVars1) ->
	        Ok = yes
	    ;   Ok = no
	    ),
	    user_time(Time1),
	    Time is Time1 - Time0,
	    format(top_level_output, '~N~n', []),
	    (   Time = 0 ->
	        true
	    ;   format(top_level_output, '(~d ms) ', [Time])
	    ),
	    format(top_level_output, '~a~n', [Ok]),
	    fail
	).




'$write_indicator' :-
	g_read('$debug_mode', DebugMode),
	'$dbg_indicator'(DebugMode, A),
	'$sys_var_read'(10, Level),
	(   Level > 1 ->
	    Level1 is Level - 1,
	    (   A = '' ->
	        format(top_level_output, '{~d}~n', [Level1])
	    ;   format(top_level_output, '{~a,~d}~n', [A, Level1])
	    )
	;   A = '' ->
	    true
	;   format(top_level_output, '{~a}~n', [A])
	),
	fail.

'$write_indicator'.




'$dbg_indicator'(trace, trace).

'$dbg_indicator'(debug, debug).

'$dbg_indicator'(nodebug, '').




'$read_query'(X, QueryVars) :-
	'$sys_var_read'(10, 1), % comment to execute in nested top-levels
	g_read('$cmd_line_query_goal', [Goal|LGoal]),
	g_assign('$cmd_line_query_goal', LGoal),
	!,
	Prompt = '| ?- ',
	(   '$sys_var_read'(12, 1) ->
	    write(top_level_output, Prompt)
	;   true),
	format(top_level_output, '~a.~n', [Goal]),
	read_term_from_atom(Goal, X, [end_of_term(eof), variable_names(QueryVars)]).

'$read_query'(X, QueryVars) :-
	read_term(top_level_input, X, [variable_names(QueryVars)]).




'$exec_query'(X, QueryVars) :-
	g_read('$debug_mode', DebugMode),
	g_assign('$debug_next', DebugMode),
	g_assign('$all_solutions', f),
	'$get_current_B'(B),
	'$call_c'('Pl_Save_Regs_For_Signal'),  % save some registers in case of CTRL+C
	'$call'(X, top_level, 0, true),
	'$call_c'('Pl_Save_Regs_For_Signal'),  % save some registers in case of CTRL+C
	'$get_current_B'(B1),
	format(top_level_output, '~N', []),
	'$catch_internal'('$set_query_vars_names'(QueryVars, ToDispVars), Err, throw('$post_query_exception'(Err)), false),
%	'$set_query_vars_names'(QueryVars, ToDispVars),
	(   fail,                             % do not activate 'alt if vars'
	    ToDispVars = [] ->
	    true                              % no alt if only anonymous vars
	;
	    '$write_solution'(ToDispVars, B1, B),
	    (   B1 > B ->
	        g_read('$all_solutions', f),          % fail for previous 'a'
	        write(top_level_output, ' ? '),
	        '$read_return'                        % fail for ';' ' ' and  'a'
	    ;   true
	    )
	).




'$set_query_vars_names'(QueryVars, ToDispVars1) :-
	name_query_vars(QueryVars, ToDispVars),
	'$remove_underscore_vars'(ToDispVars, ToDispVars1),
	name_singleton_vars(ToDispVars1),
	bind_variables(ToDispVars1, [exclude(QueryVars), namevars]).




'$remove_underscore_vars'([], []).

'$remove_underscore_vars'([Name = Term|ToDispVars], ToDispVars1) :-
	(   sub_atom(Name, 0, 1, _, '_')
	;   Term = '$VARNAME'(Name1),
	    sub_atom(Name1, 0, 1, _, '_')),
	!,
	'$remove_underscore_vars'(ToDispVars, ToDispVars1).

'$remove_underscore_vars'([X|ToDispVars], [X|ToDispVars1]) :-
	'$remove_underscore_vars'(ToDispVars, ToDispVars1).




'$write_solution'([], B1, B) :-
	!,
	(   B1 > B ->
	    format(top_level_output, '~ntrue', [])
	;   true
	).

'$write_solution'(ToDispVars, _, _) :-
	( current_op(Prior, xfx, =) ; Prior = 700), !,
	Prior1 is Prior - 1,
	'$write_solution1'(ToDispVars, Prior1).


'$write_solution1'([], _).

'$write_solution1'([Name = Value|ToDispVars], Prior) :-
	(   acyclic_term(Value) ->
	    format(top_level_output, '~n~a = ', [Name]),
	    write_term(top_level_output, Value, [quoted(true), numbervars(false), namevars(true), priority(Prior)])
	;
	    format(top_level_output, '~ncannot display cyclic term for ~a', [Name])
	),
	'$write_solution1'(ToDispVars, Prior).




'$read_return' :-
	flush_output(top_level_output),
	get_key(top_level_input, X),
	'$read_return'(X), !.




'$read_return'(10).		% newline

'$read_return'(13).		% carriage-return

'$read_return'(97) :-		% 'a'
	g_assign('$all_solutions', t), !,
	fail.

'$read_return'(59) :-		% ';'
	format(top_level_output, '~N', []), !,
	fail.

'$read_return'(32) :-		% ' ' (simulate a ';')
	format(top_level_output, '\b;~N', []), !,
	fail.

'$read_return'(_) :-
	nl(top_level_output),
	write(top_level_output, 'Action (; for next solution, a for all solutions, RET to stop) ? '),
	'$read_return'.



/* interface with command-line option consulting files */

'$exec_cmd_line_consult_files'([File|_LFile]) :-
	'$catch_internal'('$consult2'(File), error(Err, _), true, false),
	nonvar(Err),
	format('~Nwarning: command-line consulting file ~q failed due to ~q~n', [File, Err]),
	fail.
	
'$exec_cmd_line_consult_files'([_|LFile]) :-
	!,
	'$exec_cmd_line_consult_files'(LFile).

'$exec_cmd_line_consult_files'(_).		% can be another term than []


/* interface with command-line options executing goals */

'$exec_cmd_line_entry_goals'([Goal|LGoal]):-
	!,
	'$exec_cmd_line_goal'(Goal),
	'$exec_cmd_line_entry_goals'(LGoal).

'$exec_cmd_line_entry_goals'(_).		% can be another term than []




'$exec_cmd_line_goal'(Goal) :-		% called by top_level.c
	(   '$catch'('$exec_cmd1'(Goal), Err, '$exec_cmd_err'(Goal, Err), 'command-line', -1, false) ->
	    true
	;   format('~Nwarning: command-line goal ~q failed~n', [Goal])).


'$exec_cmd1'(Goal) :-
	read_term_from_atom(Goal, TermGoal, [end_of_term(eof)]),
	'$call'(TermGoal, 'command-line', -1, false).


'$exec_cmd_err'(Goal, Err) :-
	format('~Nwarning: command-line goal ~q caused exception: ~q~n', [Goal, Err]).

File: src/TopComp/top_level.c
/*-------------------------------------------------------------------------*
 * GNU Prolog                                                              *
 *                                                                         *
 * Part  : Prolog Top-level                                                *
 * File  : top_level.c                                                     *
 * Descr.: top-level command-line option checking                          *
 * Author: Daniel Diaz                                                     *
 *                                                                         *
 * Copyright (C) 1999-2013 Daniel Diaz                                     *
 *                                                                         *
 * This file is part of GNU Prolog                                         *
 *                                                                         *
 * GNU Prolog is free software: you can redistribute it and/or             *
 * modify it under the terms of either:                                    *
 *                                                                         *
 *   - the GNU Lesser General Public License as published by the Free      *
 *     Software Foundation; either version 3 of the License, or (at your   *
 *     option) any later version.                                          *
 *                                                                         *
 * or                                                                      *
 *                                                                         *
 *   - the GNU General Public License as published by the Free             *
 *     Software Foundation; either version 2 of the License, or (at your   *
 *     option) any later version.                                          *
 *                                                                         *
 * or both in parallel, as here.                                           *
 *                                                                         *
 * GNU Prolog is distributed in the hope that it will be useful,           *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU       *
 * General Public License for more details.                                *
 *                                                                         *
 * You should have received copies of the GNU General Public License and   *
 * the GNU Lesser General Public License along with this program.  If      *
 * not, see http://www.gnu.org/licenses/.                                  *
 *-------------------------------------------------------------------------*/


#include <stdio.h>
#include <string.h>

#include "../EnginePl/engine_pl.h"
#include "../BipsPl/c_supp.h"
#include "../BipsPl/inl_protos.h"
#include "copying.c"





/*---------------------------------*
 * Constants                       *
 *---------------------------------*/

/*---------------------------------*
 * Type Definitions                *
 *---------------------------------*/

/*---------------------------------*
 * Global Variables                *
 *---------------------------------*/

/*---------------------------------*
 * Function Prototypes             *
 *---------------------------------*/

static void Display_Help(void);

#define Check_Arg(i, str)  (strncmp(argv[i], str, strlen(argv[i])) == 0)


#define EXEC_CMD_LINE_GOAL  X1_24657865635F636D645F6C696E655F676F616C
#define PREDICATE_TOP_LEVEL X0_top_level

Prolog_Prototype(PREDICATE_TOP_LEVEL, 0);
Prolog_Prototype(EXEC_CMD_LINE_GOAL, 1);


/*-------------------------------------------------------------------------*
 * To define a top_level simply compile an empty source file (Prolog or C) *
 * (linking the Prolog top-level is done by default).                      *
 * This file is because we want to take into account some options/arguments*
 *-------------------------------------------------------------------------*/


/*-------------------------------------------------------------------------*
 * MAIN                                                                    *
 *                                                                         *
 * See comments in EnginePl/main.c about the use of the wrapper function.  *
 *-------------------------------------------------------------------------*/

static int
Main_Wrapper(int argc, char *argv[])
{
  int i;
  int new_argc = 0;
  char **new_argv;
  WamWord *entry_goal;
  int nb_entry_goal = 0;
  WamWord *consult_file;
  int nb_consult_file = 0;
  WamWord *query_goal;
  int nb_query_goal = 0;
  WamWord word;


  Pl_Start_Prolog(argc, argv);	/* argc and argv will be changed */

  new_argv = (char **) Malloc(sizeof(char *) * (argc + 1));
  new_argv[new_argc++] = argv[0];

  consult_file = (WamWord *) Malloc(sizeof(WamWord) * argc);
  entry_goal = (WamWord *) Malloc(sizeof(WamWord) * argc);
  query_goal = (WamWord *) Malloc(sizeof(WamWord) * argc);

  for (i = 1; i < argc; i++)
    {
      if (*argv[i] == '-' && argv[i][1] != '\0')
	{
	  if (strcmp(argv[i], "--") == 0)
	    {
	      i++;
	      break;
	    }

	  if (Check_Arg(i, "--version"))
	    {
	      Display_Copying("Prolog top-Level");
	      exit(0);
	    }

	  if (Check_Arg(i, "--init-goal"))
	    {
	      if (++i >= argc)
		Pl_Fatal_Error("Goal missing after --init-goal option");

	      A(0) = Tag_ATM(Pl_Create_Atom(argv[i]));
	      Pl_Call_Prolog(Prolog_Predicate(EXEC_CMD_LINE_GOAL, 1));
	      Pl_Reset_Prolog();
	      continue;
	    }

	  if (Check_Arg(i, "--consult-file"))
	    {
	      if (++i >= argc)
		Pl_Fatal_Error("File missing after --consult-file option");

	      consult_file[nb_consult_file++] = Tag_ATM(Pl_Create_Atom(argv[i]));
	      continue;
	    }

	  if (Check_Arg(i, "--entry-goal"))
	    {
	      if (++i >= argc)
		Pl_Fatal_Error("Goal missing after --entry-goal option");

	      entry_goal[nb_entry_goal++] = Tag_ATM(Pl_Create_Atom(argv[i]));
	      continue;
	    }

	  if (Check_Arg(i, "--query-goal"))
	    {
	      if (++i >= argc)
		Pl_Fatal_Error("Goal missing after --query-goal option");

	      query_goal[nb_query_goal++] = Tag_ATM(Pl_Create_Atom(argv[i]));
	      continue;
	    }

	  if (Check_Arg(i, "-h") || Check_Arg(i, "--help"))
	    {
	      Display_Help();
	      exit(0);
	    }
	}
      /* unknown option is simply ignored (passed to Prolog) */
      new_argv[new_argc++] = argv[i];
    }

  while(i < argc)
    new_argv[new_argc++] = argv[i++];

  new_argv[new_argc] = NULL;

  pl_os_argc = new_argc;
  pl_os_argv = new_argv;

  if (nb_consult_file)
    {
      word = Pl_Mk_Proper_List(nb_consult_file, consult_file);
      Pl_Blt_G_Assign(Tag_ATM(Pl_Create_Atom("$cmd_line_consult_file")), word);
    }
  Free(consult_file);

  if (nb_entry_goal)
    {
      word = Pl_Mk_Proper_List(nb_entry_goal, entry_goal);
      Pl_Blt_G_Assign(Tag_ATM(Pl_Create_Atom("$cmd_line_entry_goal")), word);
    }
  Free(entry_goal);

  if (nb_query_goal)
    {
      word = Pl_Mk_Proper_List(nb_query_goal, query_goal);
      Pl_Blt_G_Assign(Tag_ATM(Pl_Create_Atom("$cmd_line_query_goal")), word);
    }
  Free(query_goal);

  Pl_Reset_Prolog();
  Pl_Call_Prolog(Prolog_Predicate(PREDICATE_TOP_LEVEL, 0));

  Pl_Stop_Prolog();
  return 0;
}

int
main(int argc, char *argv[])
{
  return Main_Wrapper(argc, argv);
}


/*-------------------------------------------------------------------------*
 * DISPLAY_HELP                                                            *
 *                                                                         *
 *-------------------------------------------------------------------------*/
static void
Display_Help(void)
#define L(msg)  fprintf(stderr, "%s\n", msg)
{
  fprintf(stderr, "Usage: %s [OPTION]... \n", TOP_LEVEL);
  L("");
  L("  --consult-file FILE         consult FILE inside the the top-level");
  L("  --init-goal    GOAL         execute GOAL before entering the top-level");
  L("  --entry-goal   GOAL         execute GOAL inside the top-level");
  L("  --query-goal   GOAL         execute GOAL as a query for the top-level");
  L("  -h, --help                  print this help and exit");
  L("  --version                   print version number and exit");
  L("  --                          do not parse the rest of the command-line");
  L("");
  L("Report bugs to bug-prolog@gnu.org.");
}

#undef L
